lib/type/param_string_struct.ex

```diff
@@ -0,0 +1,282 @@
+
+  ```elixir
+  defmodule MyQueryParams2 do
+    use Antikythera.ParamStringStruct,
+      fields: [
+        q: Croma.TypeGen.nilable(Croma.String),
+        date: {Date, [default: ~D[1970-01-01]]}
+      ]
+  end
+
+  plug Antikythera.Plug.ParamsValidator, :validate, query_params: MyQueryParams2
+  ```
+
+  In the example above, the request without any query parameters is allowed; `q` is set to `nil`, and `date` is set to `~D[1970-01-01]`.
+
+  ### Naming convention
+
+  By default, the field name in the struct is the same as the parameter key.
+  You can specify a different key name scheme using the `:accept_case` option, which is the same as that of `Croma.Struct`.
+
+  ```elixir
+  defmodule MyQueryParams3 do
+    use Antikythera.ParamStringStruct,
+      accept_case: :lower_camel,
+      fields: [
+        item_id: Croma.PosInteger  # The parameter key name "itemId" is also accepted.
+      ]
+  end
+  ```
+
+  ## Limitations
+
+  The preprocessor must be a captured named function like `&Module.function/arity` to be used in module attributes internally.
+  """
+
+  defmodule Preprocessor do
+    @moduledoc """
+    Preprocessors for string parameters.
+    """
+
+    @type t :: (nil | String.t() -> Croma.Result.t() | term)
+
+    @doc false
+    defun default(mod :: v[module]) :: Croma.Result.t(t()) do
+      # The default preprocessors are defined as a capture form `&Mod.fun/arity`, which can be used in module attributes.
+      case Module.split(mod) do
+        # Preprocessors for Croma built-in types
+        ["Croma", "Boolean"] ->
+          {:ok, &__MODULE__.to_boolean/1}
+
+        ["Croma", "Float"] ->
+          {:ok, &String.to_float/1}
+
+        ["Croma", "Integer"] ->
+          {:ok, &String.to_integer/1}
+
+        ["Croma", "NegInteger"] ->
+          {:ok, &String.to_integer/1}
+
+        ["Croma", "NonNegInteger"] ->
+          {:ok, &String.to_integer/1}
+
+        ["Croma", "Number"] ->
+          {:ok, &__MODULE__.to_number/1}
+
+        ["Croma", "PosInteger"] ->
+          {:ok, &String.to_integer/1}
+
+        ["Croma", "String"] ->
+          {:ok, &__MODULE__.passthrough_string/1}
+
+        # Preprocessors for DateTime-related types
+        ["Date"] ->
+          {:ok, &Date.from_iso8601/1}
+
+        ["DateTime"] ->
+          {:ok, &__MODULE__.to_datetime/1}
+
+        ["NaiveDateTime"] ->
+          {:ok, &NaiveDateTime.from_iso8601/1}
+
+        ["Time"] ->
+          {:ok, &Time.from_iso8601/1}
+
+        # Preprocessors for nilable types
+        ["Croma", "TypeGen", "Nilable" | original_mod_split] ->
+          original_mod = Module.safe_concat(original_mod_split)
+
+          original_mod
+          |> default()
+          |> Croma.Result.map(&generate_nilable_preprocessor(&1, original_mod))
+
+        _ ->
+          {:error, :no_default_preprocessor}
+      end
+    end
+
+    @doc """
+    Converts a string to a boolean value.
+    """
+    defun to_boolean(s :: nil | String.t()) :: boolean do
```

Comment:
Do you want gear developers to use these `defun` functions other than `ParamStringStruct` ? If so, I think you should create a new module for them like [`MapUtil`](https://github.com/access-company/antikythera/blob/master/lib/util/map_util.ex)

Reply:
If you move them to an other module, you can change `@moduledoc false` and make `Preprocessor` implementation detail.

Reply:
Added `Antikythera.StringPreprocessor` module and hid the preprocessor module at 88c9f5e.
Also hid `AntikytheraCore.BodyJsonCommon.PreprocessorGenerator` at 2a4c8db.